File: backend/src/main/java/com/joying/chat/controller/ChatRoomController.java
package com.joying.chat.controller;

import java.util.Arrays;
import java.util.List;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import com.joying.chat.dto.ChatRoomResponse;
import com.joying.chat.dto.ChatRoomSettingsResponse;
import com.joying.chat.dto.CreateChatRoomRequest;
import com.joying.chat.dto.UpdateChatRoomSettingsRequest;
import com.joying.chat.service.ChatRoomService;
import com.joying.chat.service.ChatService;
import com.joying.common.response.ApiResponse;

import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import lombok.RequiredArgsConstructor;

/**
 * 채팅방을 만들고 목록과 상세를 준다.
 */
@Tag(name = "Chat Room", description = "채팅방 API")
@RestController
@RequestMapping("/api/v1/chat-rooms")
@RequiredArgsConstructor
public class ChatRoomController {

	private static final Logger log = LoggerFactory.getLogger(ChatRoomController.class);

	private final ChatRoomService chatRoomService;
	private final ChatService chatService;

	@Operation(summary = "채팅방 생성 또는 조회",
		description = "상품에 대한 채팅방을 만들거나 이미 있으면 그것을 돌려준다")
	@PostMapping
	public ResponseEntity<ApiResponse.SuccessBody<ChatRoomResponse>> createOrGetChatRoom(
		@RequestBody CreateChatRoomRequest request) {

		Long memberId = CurrentMember.id();
		log.info("채팅방 생성/조회 요청: productId={}, memberId={}", request.getProductId(), memberId);

		// 지연 로딩이 트랜잭션 밖에서 끊기지 않도록 서비스에서 응답까지 만들어 온다
		ChatRoomResponse chatRoom =
			chatRoomService.getOrCreateChatRoomResponse(request.getProductId(), memberId);
		return ApiResponse.ok("채팅방이 생성되었습니다", chatRoom);
	}

	@Operation(summary = "내 채팅방 목록 조회",
		description = "참여 중인 방을 돌려준다. 총 안읽음 개수는 X-Total-Unread-Count 헤더에 담는다")
	@GetMapping
	public ResponseEntity<ApiResponse.SuccessBody<List<ChatRoomResponse>>> getMyChatRooms(
		@RequestParam(required = false) String include) {

		Long memberId = CurrentMember.id();
		boolean includeMember = wants(include, "member");

		List<ChatRoomResponse> chatRooms = chatRoomService.getMyChatRooms(memberId, includeMember);

		// 배지를 그리는 데만 쓰므로 본문이 아니라 헤더로 보낸다
		long totalUnreadCount = chatRooms.stream()
			.mapToLong(ChatRoomResponse::getUnreadCount)
			.sum();

		log.info("채팅방 목록 조회 완료: memberId={}, 방 개수={}, 총 안읽음={}",
			memberId, chatRooms.size(), totalUnreadCount);

		return ResponseEntity.ok()
			.header("X-Total-Unread-Count", String.valueOf(totalUnreadCount))
			.body(ApiResponse.ok("채팅방 목록 조회 완료", chatRooms).getBody());
	}

	@Operation(summary = "채팅방 상세 조회")
	@GetMapping("/{chatRoomId}")
	public ResponseEntity<ApiResponse.SuccessBody<ChatRoomResponse>> getChatRoomDetail(
		@PathVariable Long chatRoomId, @RequestParam(required = false) String include) {

		Long memberId = CurrentMember.id();
		ChatRoomResponse chatRoom =
			chatRoomService.getChatRoomDetail(chatRoomId, memberId, wants(include, "member"));
		return ApiResponse.ok("채팅방 상세 조회 완료", chatRoom);
	}

	@Operation(summary = "채팅방 나가기",
		description = "이 사람만 방에서 빠진다. 방 자체는 남고 상대는 계속 볼 수 있다")
	@DeleteMapping("/{chatRoomId}")
	public ResponseEntity<Void> leaveChatRoom(@PathVariable Long chatRoomId) {
		Long memberId = CurrentMember.id();
		log.info("채팅방 나가기 요청: chatRoomId={}, memberId={}", chatRoomId, memberId);

		chatRoomService.leaveChatRoom(chatRoomId, memberId);
		return ApiResponse.noContent();
	}

	@Operation(summary = "채팅방 설정 업데이트",
		description = "고정과 알림 설정을 바꾼다. 보내지 않은 것은 바꾸지 않는다")
	@PatchMapping("/{chatRoomId}/settings")
	public ResponseEntity<ApiResponse.SuccessBody<ChatRoomSettingsResponse>> updateSettings(
		@PathVariable Long chatRoomId, @RequestBody UpdateChatRoomSettingsRequest request) {

		Long memberId = CurrentMember.id();
		ChatRoomSettingsResponse settings = chatService.updateSettings(
			chatRoomId, memberId, request.getIsPinned(), request.getIsMuted());
		return ApiResponse.ok("채팅방 설정이 업데이트되었습니다", settings);
	}

	/**
	 * {@code include=member,other} 처럼 쉼표로 붙여 보내는 값을 가린다.
	 */
	private boolean wants(String include, String what) {
		if (include == null) {
			return false;
		}
		return Arrays.stream(include.split(","))
			.map(String::trim)
			.anyMatch(what::equals);
	}
}

File: backend/src/main/kotlin/com/joying/chat/controller/ChatRoomController.kt
package com.joying.chat.controller

import com.joying.chat.dto.ChatRoomResponse
import com.joying.chat.dto.ChatRoomSettingsResponse
import com.joying.chat.dto.CreateChatRoomRequest
import com.joying.chat.dto.UpdateChatRoomSettingsRequest
import com.joying.chat.service.ChatRoomService
import com.joying.chat.service.ChatService
import com.joying.common.response.ApiResponse
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.tags.Tag
import org.slf4j.LoggerFactory
import org.springframework.http.ResponseEntity
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.web.bind.annotation.*

/**
 * 채팅방 REST API Controller
 *
 * 채팅방 생성, 조회, 관리 기능
 */
@Tag(name = "Chat Room", description = "채팅방 API")
@RestController
@RequestMapping("/api/v1/chat-rooms")
class ChatRoomController(
    private val chatRoomService: ChatRoomService,
    private val chatService: ChatService
) {
    private val logger = LoggerFactory.getLogger(ChatRoomController::class.java)

    /**
     * 채팅방 생성 또는 조회
     *
     * 이미 존재하는 채팅방이 있으면 기존 채팅방 반환
     * (상품 + 구매자 + 판매자 조합으로 유일성 보장)
     *
     * @param request 채팅방 생성 요청 (productId)
     * @return 생성/조회된 채팅방 정보
     */
    @Operation(summary = "채팅방 생성 또는 조회", description = "상품에 대한 채팅방을 생성하거나 기존 채팅방을 반환합니다")
    @PostMapping
    fun createOrGetChatRoom(
        @RequestBody request: CreateChatRoomRequest
    ): ResponseEntity<ApiResponse.SuccessBody<ChatRoomResponse>> {
        val memberId = getCurrentMemberId()

        logger.info("채팅방 생성/조회 요청: productId={}, memberId={}", request.productId, memberId)

        // Service에서 DTO까지 완성해서 반환 (lazy loading 문제 방지)
        val chatRoomDto = chatRoomService.getOrCreateChatRoomResponse(request.productId, memberId)

        return ApiResponse.ok("채팅방이 생성되었습니다", chatRoomDto)
    }

    /**
     * 내 채팅방 목록 조회
     *
     * 내가 참여 중인 모든 채팅방 목록 반환
     * (안읽은 메시지 개수 포함, 응답 헤더로 총 안읽은 개수 제공)
     *
     * @return 채팅방 목록 (+ 헤더: X-Total-Unread-Count)
     */
    @Operation(
        summary = "내 채팅방 목록 조회",
        description = "내가 참여 중인 모든 채팅방 목록을 반환합니다. 응답 헤더 X-Total-Unread-Count에 총 안읽은 메시지 개수가 포함됩니다."
    )
    @GetMapping
    suspend fun getMyChatRooms(): ResponseEntity<ApiResponse.SuccessBody<List<ChatRoomResponse>>> {
        val memberId = getCurrentMemberId()

        logger.info("채팅방 목록 조회 요청: memberId={}", memberId)

        val chatRooms = chatRoomService.getMyChatRooms(memberId)

        // 총 안읽은 개수 계산 (헤더로 전달)
        val totalUnreadCount = chatRooms.sumOf { it.unreadCount }

        val response = ApiResponse.ok("채팅방 목록 조회 완료", chatRooms)

        return ResponseEntity.ok()
            .header("X-Total-Unread-Count", totalUnreadCount.toString())
            .body(response.body)
    }

    /**
     * 채팅방 상세 조회
     *
     * @param chatRoomId 채팅방 ID
     * @param include 포함할 추가 정보 (member: 참여자 온라인 상태 등)
     * @return 채팅방 상세 정보
     */
    @Operation(
        summary = "채팅방 상세 조회",
        description = "특정 채팅방의 상세 정보를 반환합니다. include=member 파라미터로 참여자 온라인 상태를 포함할 수 있습니다."
    )
    @GetMapping("/{chatRoomId}")
    suspend fun getChatRoomDetail(
        @PathVariable chatRoomId: Long,
        @RequestParam(required = false) include: String?
    ): ResponseEntity<ApiResponse.SuccessBody<ChatRoomResponse>> {
        val memberId = getCurrentMemberId()

        logger.info("채팅방 상세 조회 요청: chatRoomId={}, memberId={}, include={}", chatRoomId, memberId, include)

        // include 파라미터 파싱 (쉼표로 구분된 값)
        val includes = include?.split(",")?.map { it.trim() } ?: emptyList()
        val includeMember = includes.contains("member")

        val chatRoom = chatRoomService.getChatRoomDetail(chatRoomId, memberId, includeMember)

        return ApiResponse.ok("채팅방 상세 조회 완료", chatRoom)
    }

    /**
     * 채팅방 나가기
     *
     * 채팅방 상태를 CLOSED로 변경
     *
     * @param chatRoomId 채팅방 ID
     * @return 성공 메시지
     */
    @Operation(summary = "채팅방 나가기", description = "채팅방을 나가고 상태를 종료로 변경합니다")
    @DeleteMapping("/{chatRoomId}")
    fun leaveChatRoom(
        @PathVariable chatRoomId: Long
    ): ResponseEntity<Void> {
        val memberId = getCurrentMemberId()

        logger.info("채팅방 나가기 요청: chatRoomId={}, memberId={}", chatRoomId, memberId)

        chatRoomService.leaveChatRoom(chatRoomId, memberId)

        return ApiResponse.noContent()
    }

    /**
     * 채팅방 설정 업데이트 (통합)
     *
     * 고정/알림 설정을 한번에 업데이트합니다.
     * null로 전달하면 해당 설정은 변경하지 않습니다.
     *
     * @param chatRoomId 채팅방 ID
     * @param request 업데이트할 설정 (isPinned, isMuted)
     * @return 업데이트된 설정
     */
    @Operation(
        summary = "채팅방 설정 업데이트",
        description = "채팅방 고정/알림 설정을 업데이트합니다. null로 전달하면 해당 설정은 변경하지 않습니다."
    )
    @PatchMapping("/{chatRoomId}/settings")
    fun updateSettings(
        @PathVariable chatRoomId: Long,
        @RequestBody request: UpdateChatRoomSettingsRequest
    ): ResponseEntity<ApiResponse.SuccessBody<ChatRoomSettingsResponse>> {
        val memberId = getCurrentMemberId()

        logger.info(
            "채팅방 설정 업데이트 요청: chatRoomId={}, memberId={}, isPinned={}, isMuted={}",
            chatRoomId,
            memberId,
            request.isPinned,
            request.isMuted
        )

        val settings = chatService.updateSettings(chatRoomId, memberId, request.isPinned, request.isMuted)

        return ApiResponse.ok("채팅방 설정이 업데이트되었습니다", settings)
    }

    /**
     * SecurityContext에서 현재 인증된 사용자 ID 반환
     *
     * JwtAuthenticationFilter가 쿠키 또는 Authorization 헤더에서
     * JWT 토큰을 추출하고 SecurityContext에 인증 정보를 설정합니다.
     *
     * @return 사용자 ID
     * @throws IllegalStateException 인증되지 않은 경우
     */
    private fun getCurrentMemberId(): Long {
        val authentication = SecurityContextHolder.getContext().authentication
            ?: throw IllegalStateException("인증 정보가 없습니다")

        return authentication.name.toLong()
    }
}